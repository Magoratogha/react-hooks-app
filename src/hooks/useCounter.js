import { useState } from "react";

function useCounter(initialCounter = 0) {
    const [counter, setCounter] = useState(initialCounter);

    const increment = (factor = 1) => {
        setCounter(counter + factor);
    };

    const decrement = (factor = 1) => {
        setCounter(counter - factor);
    };

    const reset = () => {
        setCounter(initialCounter);
    };

    return {
        counter,
        increment,
        decrement,
        reset
    };
}

export default useCounter;