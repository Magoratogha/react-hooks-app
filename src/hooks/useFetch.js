import { useEffect, useRef, useState } from "react";

function useFetch(url) {

    const isMounted = useRef(true);
    const [state, setState] = useState({
        data: null,
        isLoading: true,
        error: null
    });

    useEffect(() => {
        return () => {
            isMounted.current = false;
        }
    }, [])


    useEffect(() => {
        setState({
            data: null,
            isLoading: true,
            error: null
        });

        fetch(url)
            .then((response) => response.json())
            .then((data) => {
                if (isMounted.current) {
                    setState({ data, isLoading: false, error: null });
                }
            })
            .catch((error) => setState({ data: null, isLoading: false, error }));
    }, [url]);

    return state;
}

export default useFetch;