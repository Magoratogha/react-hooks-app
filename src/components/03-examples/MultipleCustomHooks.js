import { Fragment } from "react";
import useCounter from "../../hooks/useCounter";
import useFetch from "../../hooks/useFetch";
import "./MultipleCustomHooks.css";

function MultipleCustomHooks() {

    const { counter, increment } = useCounter(1);
    const requestEndpoint = 'https://www.breakingbadapi.com/api/quotes/' + counter;
    const { isLoading, data } = useFetch(requestEndpoint);
    const { quote, author } = !!data && data[0];
    
    return (
        <Fragment>
            <h1>Breaking Bad quotes</h1>
            <hr></hr>
            { isLoading ? 
                <div className="alert alert-info text-center">Loading...</div>
                :
                <blockquote className="blockquote text-end">
                    <p className="mb-3">{ quote }</p>
                    <footer className="blockquote-footer">{ author }</footer>
                </blockquote>
            }
            <button
                onClick={ () => increment(1) }
                className="btn btn-outline-dark">
                    Next quote
                </button>
        </Fragment>
    );
}

export default MultipleCustomHooks;