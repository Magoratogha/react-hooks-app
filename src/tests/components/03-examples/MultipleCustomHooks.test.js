import { queryByAttribute, render, screen } from "@testing-library/react";
import MultipleCustomHooks from "../../../components/03-examples/MultipleCustomHooks";
import useCounter from "../../../hooks/useCounter";
import useFetch from "../../../hooks/useFetch";
jest.mock('../../../hooks/useCounter');
jest.mock('../../../hooks/useFetch');

describe('Test on <MultipleCustomHooks />', () => {

    beforeEach(() => {
        useCounter.mockReturnValue({
            counter: 10,
            increment: () => { }
        });
    });

    test('should render properly', () => {
        useFetch.mockReturnValue({
            data: null,
            isLoading: true,
            error: null
        });
        const { container } = render(<MultipleCustomHooks />);
        expect(container).toMatchSnapshot();
    });

    test('should show data when its fetched', () => {
        useFetch.mockReturnValue({
            data: [{
                author: 'Magora',
                quote: 'Hello world'
            }],
            isLoading: false,
            error: null
        });
        const { container } = render(<MultipleCustomHooks />);
        expect(queryByAttribute('class', container, 'alert', { exact: false })).toBeNull();
        expect(screen.getByText('Hello world', { selector: 'p' })).toBeTruthy();
        expect(screen.getByText('Magora', { selector: 'footer' })).toBeTruthy();
    });
});