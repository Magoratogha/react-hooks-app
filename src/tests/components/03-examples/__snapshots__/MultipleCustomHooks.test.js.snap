// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <MultipleCustomHooks /> should render properly 1`] = `
<div>
  <h1>
    Breaking Bad quotes
  </h1>
  <hr />
  <div
    class="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    class="btn btn-outline-dark"
  >
    Next quote
  </button>
</div>
`;
